drivers/mdio/mdio_esp32.c

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (c) 2022 Grant Ramsay <grant.ramsay@hotmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT espressif_esp32_mdio
+
+#include <soc.h>
+#include <zephyr/drivers/mdio.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/logging/log.h>
+
+#include <driver/periph_ctrl.h>
+#include <esp_mac.h>
+#include <hal/emac_hal.h>
+#include <hal/emac_ll.h>
+
+LOG_MODULE_REGISTER(mdio_esp32, CONFIG_MDIO_LOG_LEVEL);
+
+#define PHY_OPERATION_TIMEOUT_US 1000
+
+struct mdio_esp32_dev_data {
+	struct k_sem sem;
+	emac_hal_context_t hal;
+};
+
+struct mdio_esp32_dev_config {
+	const struct pinctrl_dev_config *pcfg;
+};
+
+static int mdio_transfer(const struct device *dev, uint8_t prtad, uint8_t devad,
+			 bool write, uint16_t data_in, uint16_t *data_out)
+{
+	struct mdio_esp32_dev_data *const dev_data = dev->data;
+
+	k_sem_take(&dev_data->sem, K_FOREVER);
+
+	if (emac_ll_is_mii_busy(dev_data->hal.mac_regs)) {
+		LOG_ERR("phy busy");
+		k_sem_give(&dev_data->sem);
+		return -EBUSY;
+	}
+
+	if (write) {
+		emac_ll_set_phy_data(dev_data->hal.mac_regs, data_in);
+	}
+	emac_hal_set_phy_cmd(&dev_data->hal, prtad, devad, write);
+
+	/* Poll until operation complete */
+	bool success = false;
+
+	for (uint32_t t_us = 0; t_us < PHY_OPERATION_TIMEOUT_US; t_us += 100) {
+		k_sleep(K_USEC(100));
+		if (!emac_ll_is_mii_busy(dev_data->hal.mac_regs)) {
+			success = true;
+			break;
+		}
+	}
+	if (!success) {
+		LOG_ERR("phy timeout");
+		k_sem_give(&dev_data->sem);
+		return -ETIMEDOUT;
+	}
+
+	if (!write && data_out != NULL) {
+		*data_out = emac_ll_get_phy_data(dev_data->hal.mac_regs);
+	}
+
+	k_sem_give(&dev_data->sem);
+
+	return 0;
+}
+
+static int mdio_esp32_read(const struct device *dev, uint8_t prtad, uint8_t devad,
+			 uint16_t *data)
+{
+	return mdio_transfer(dev, prtad, devad, false, 0, data);
+
+}
+
+static int mdio_esp32_write(const struct device *dev, uint8_t prtad,
+			  uint8_t devad, uint16_t data)
+{
+	return mdio_transfer(dev, prtad, devad, true, data, NULL);
+}
+
+static void mdio_esp32_bus_enable(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+}
+
+static void mdio_esp32_bus_disable(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+}
+
+static int mdio_esp32_initialize(const struct device *dev)
+{
+	const struct mdio_esp32_dev_config *const cfg = dev->config;
+	struct mdio_esp32_dev_data *const dev_data = dev->data;
+	int res;
+
+	k_sem_init(&dev_data->sem, 1, 1);
```

Comment:
same here, is this a binary semaphore? if so it should be initialized to 1 key and 0 keys available

Reply:
The other semaphore (eth_esp32.c) was incorrectly initialized. However I believe this one is correct.
In this case there is one device/resource, and it is ready to be used/taken after initialization

Reply:
Yes now I see it, just a minor question (feel free to resolve), why not use a mutex instead? Do you believe the priority inversion risk here is negligible?

Reply:
It could be a mutex, TBH I used this because it was similar to how the SAM MDIO driver did it (mdio_sam.c).
PHY operations are pretty fast (<1ms) so shouldn't block for very long, and most are likely done from a single thread (except MDIO_SHELL which seems more for debug)